src/embeddings/cooccurrence_counter.cpp

```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    friend class cooccurrence_counter;
+
+    cooccurrence_counter* counter_;
+    const std::size_t max_bytes_;
+    map_t cooccurrences_;
+    std::unique_ptr<analyzers::token_stream> stream_;
+};
+
+namespace
+{
+hashing::probe_map<std::string, uint64_t>
+load_vocab(const std::string& filename)
+{
+    using map_type = hashing::probe_map<std::string, uint64_t>;
+
+    std::ifstream input{filename, std::ios::binary};
+    auto size = io::packed::read<uint64_t>(input);
+    auto reserve_size = static_cast<std::size_t>(
+        std::ceil(size / map_type::default_max_load_factor()));
+
+    printing::progress progress{" > Loading vocab: ", size};
+    map_type vocab{reserve_size};
+    for (uint64_t tid{0}; tid < size; ++tid)
+    {
+        progress(tid);
+        auto word = io::packed::read<std::string>(input);
+        io::packed::read<uint64_t>(input);
+
+        vocab[word] = tid;
+    }
+
+    return vocab;
+}
+}
+
+cooccurrence_counter::cooccurrence_counter(configuration config,
+                                           parallel::thread_pool& pool)
+    : prefix_{std::move(config.prefix)},
+      max_ram_{config.max_ram},
+      merge_fanout_{config.merge_fanout},
+      window_size_{config.window_size},
+      break_on_tags_{config.break_on_tags},
+      vocab_{load_vocab(prefix_ + "/vocab.bin")},
+      pool_(pool)
+{
+    LOG(info) << "Loaded vocabulary of size " << vocab_.size() << " occupying "
+              << printing::bytes_to_units(vocab_.bytes_used()) << ENDLG;
+
+    if (vocab_.bytes_used() > max_ram_)
+        throw cooccurrence_exception{"RAM limit too restrictive"};
+    max_ram_ -= vocab_.bytes_used();
+}
+
+void cooccurrence_counter::count(corpus::corpus& docs,
+                                 const analyzers::token_stream& stream)
+{
+    if (chunk_num_ != 0)
+        throw cooccurrence_exception{
+            "cooccurrence_counters may not be re-used"};
+
+    num_tokenizing_ = pool_.size();
+    printing::progress progress{" > Counting cooccurrences: ", docs.size()};
+    corpus::parallel_consume(
+        docs, pool_,
+        [&]() {
+            return cooccurrence_buffer{this, max_ram_ / pool_.size(), stream};
+        },
+        [&](cooccurrence_buffer& buffer, const corpus::document& doc) {
+            {
+                std::lock_guard<std::mutex> lock{io_mutex_};
+                progress(doc.id());
+            }
+
+            buffer.stream_->set_content(analyzers::get_content(doc));
+
+            std::deque<uint64_t> history;
+            while (*buffer.stream_)
+            {
+                auto tok = buffer.stream_->next();
+
+                if (tok == "<s>" && break_on_tags_)
+                {
+                    history.clear();
+                }
+                else if (tok == "</s>" && break_on_tags_)
+                {
+                    continue;
+                }
+                else
+                {
+                    // ignore out-of-vocabulary words
```

Comment:
Levy and Goldberg have this to say:

> **Deleting  Rare  Words  (`del`)** While  it  is  common to ignore words that are rare in the training corpus, word2vec removes  these  tokens  from the corpus before creating context windows.   As with subsampling, this variation narrows the distance between tokens, inserting new word-context pairs that did not exist in the original corpus with the same window size. Though this variation may also  have  an  effect  on  performance,  preliminary experiments  showed  that  it  was  small,  and  we therefore do not investigate its effect in this paper.

Not sure if I care about adding it. It adds a bunch of if/else into this loop to handle the case where `tid` and/or `*it` is not actually a value in the vocabulary.